epytope/IO/MartsAdapter.py

```diff
@@ -7,58 +7,251 @@
+        :param str value: The value of the filter
+        :param str name: The name of the XML field (default: Attribute)
+
+        :return: The SubElement with attribute
+        :rtype: ElementTree.SubElement
+        """
+        element = ElementTree.SubElement(parent, name)
+        element.set(filter_id, id)
+        element.set(filter_value, value)
+        return element
+
+    def __search_for_resources(self, xml_root):
+        """
+        Perform search (GET) using the given XML as query value
+
+        :param ElementTree Element xml_root: The XML of the query
+
+        :return: The found resources as data frame
+        :rtype: pandas.core.frame.DataFrame
+        """
+        f = io.BytesIO()
+        et = ElementTree.ElementTree(xml_root)
+        et.write(f, encoding="UTF-8", xml_declaration=True)
+        response = requests.get(self.biomart_url, params={
+                                "query": f.getvalue().decode("utf8")})
+        result = pd.read_csv(io.StringIO(
+            response.content.decode('utf-8')), delimiter='\t')
+        return result
+
+    def __chunks(self, lst, n):
+        for i in range(0, len(lst), n):
+            yield lst[i:i + n]
+
+    def list_archives(self):
+        """
+        Fetches the available Ensembl archives from the ensembl website
+
+        :return: The available Ensembl archives
+        :rtype: pandas.core.frame.DataFrame
+        """
+        html_text = requests.get(self.biomart_archive_url).text
+        soup = BeautifulSoup(html_text, 'html.parser')
+
+        archives = []
+        for archive in soup.find_all('a', string=re.compile(r'http[s]?://.*ensembl\\.org|[A-Z][a-z]{2} [0-9]{4}')):
+            url = archive['href']
+            version, date = archive.contents[0].split(':')
+            archives.append([version, url, date])
+
+        return pd.DataFrame(archives, columns=['name', 'url', 'date'])
+
+    def get_marts(self):
+        """
+        Fetches the mart registry with information on available marts
+
+        :return: The available marts for this server
+        :rtype: pandas.core.frame.DataFrame
+        """
+        registry = requests.get(self.biomart_url, params={"type": "registry"})
+        df = pd.read_xml(io.StringIO(registry.content.decode('utf-8')))
+        return df
+
+    def get_datasets(self, mart_name):
+        """
+        Fetches available datasets for a given mart name
+
+        :param str mart_name: The name of the mart
+
+        :return: The available datasets for this mart
+        :rtype: pandas.core.frame.DataFrame
+        """
+        datasets = requests.get(self.biomart_url, params={
+                                "type": "datasets", "mart": mart_name})
+        df = pd.read_csv(io.StringIO(
+            datasets.content.decode('utf-8')), delimiter='\t')
+        df_slice = df.iloc[:, [1, 2, 4, 7, 8]]
+        df_slice.dropna(how='all', inplace=True)
+        df_slice.columns = ['dataset_id', 'dataset_name',
```

Comment:
Will these fields persist for different fetched datasets?

Reply:
Do you mean the order/number of the returned columns? I would hope so... but maybe it is a good idea to cover that with an additional test. 👍

Reply:
Yep indeed, I saw you already included tests :) Nice